opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -145,6 +145,62 @@ def from_map(cls, maxlen, mapping):
 Link = namedtuple("Link", ("context", "attributes"))
 
 
+class SpanProcessor:
+    def on_start(self, span: "Span") -> None:
+        """Called when a :class:`Span` is started.
+
+        This method is called synchronously on the execution thread, therefore
+        it should not block or throw an exception.
+
+        Args:
+            span: The :class:`Span` that just started.
+        """
+
+    def on_end(self, span: "Span") -> None:
+        """Called when a :class:`Span` is ended.
+
+        This method is called synchronously on the execution thread, therefore
+        it should not block or throw an exception.
+
+        Args:
+            span: The :class:`Span` that just ended.
+        """
+
+    def shutdown(self) -> None:
+        """Called when a :class:`Tracer` is shutdown."""
+
+
+# Used when no SpanProcessor has been added to the Tracer
+_NO_OP_SPAN_PROCESSOR = SpanProcessor()
+
+
+class MultiSpanProcessor(SpanProcessor):
+    """Implementation of :class:`SpanProcessor` that forwards all received
+    events to a list of `SpanProcessor`.
+    """
+
+    def __init__(self):
+        # use a tuple to avoid race conditions when adding a new span and
+        # interating throw it on "on_start" and "on_end".
+        self._span_processors = ()
+
+    def add_span_processor(self, span_processor: SpanProcessor):
+        """Adds a SpanProcessor to the list handled by this instance."""
+        self._span_processors = self._span_processors + (span_processor,)
```

Comment:
I believe yes for CPython at least.

Reply:
Hmm, but still the access that reads the _span_processors and the one that writes it are not atomic together, so this will still need some kind of locking (or, preferably IMHO, as this would sacrifice performance for an edge case), we document that this method is not thread safe (and recommend it is only called during an initial initialization phase or some such)

Reply:
Having thread safe solution without using locks in the reader side looks complicated to me, so I agree that we can assume this function is not thread safe and document it.

Reply:
We could have a single, simple lock for the SDK `Tracer` itself. In Java we have this one for both `shutdown()` and `addSpanProcessor()` (and given the fact we will have to add `shutdown()`, I see no reason to not add this now).

Observe that a lock here shouldn't be much of a problem, as it will *not* impact `Span` (or related) operations ;)

Reply:
Hmm true, when we assume atomic assignment, we won't need a lock when we just read the span processor list.

Reply:
If we assume atomic assignment then just a lock to protect multiple calls to `add_span_processor`  will suffice.  I don't really know if we should assume that because it could not work in all the cases, different implementations than CPython for instance.